### PR TITLE
[ie/laracasts] Add extractors

### PR DESCRIPTION
**Pull request by ASertacAkkaya**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Adds support for downloading series from Laracasts.com (using Vimeo embed).

https://laracasts.com/series

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/_extractors.py [hunk 1]
```diff
@@ -969,11 +969,16 @@
...
+from .laracasts import (
+    LaracastsIE,
+    LaracastsPlaylistIE,
+)
 from .lastfm import (
     LastFMIE,
     LastFMPlaylistIE,
     LastFMUserIE,
 )
+
```
Comment (bashonly) [inline yt_dlp/extractor/_extractors.py:981]:
```suggestion
```

File: yt_dlp/extractor/laracasts.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
+from .common import InfoExtractor
+from .vimeo import VimeoIE
+
+
+class LaracastsPlaylistIE(InfoExtractor):
+    IE_NAME = 'laracasts:series'
+    _VALID_URL = r'https?://(?:www\.)?laracasts\.com/series/(?P<series>[^/?#]+)'
```
Comment (bashonly) [inline yt_dlp/extractor/laracasts.py:7]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?laracasts\.com/series/(?P<series>[^/?#]+)/?(?:$|[?#])'
```

File: yt_dlp/extractor/laracasts.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    def _real_extract(self, url):
+        mobj = self._match_valid_url(url)
+        display_id = mobj.group('series')
+        webpage = self._download_webpage(url, display_id)
+        episode_count = int(self._search_regex(r'episodeCount&quot;:(?P<episode_count>[0-9]+)', webpage, 'episode_count'))
+        playlist_title = self._search_regex(r'title&quot;:&quot;(?P<playlist_title>[^&]+)&quot;,', webpage, 'playlist_title')
+        playlist_id = self._search_regex(r'id&quot;:(?P<playlist_id>[0-9]+)', webpage, 'playlist_id')
+        return self.playlist_result(self._entries(display_id, episode_count), playlist_id, playlist_title)
+
+class LaracastsIE(InfoExtractor):
```
Comment (bashonly) [inline yt_dlp/extractor/laracasts.py:31]:
```suggestion
        return self.playlist_result(self._entries(display_id, episode_count), playlist_id, playlist_title)


class LaracastsIE(InfoExtractor):
```

**Comment by seproDev** [comment]
I have reworked the extractor a bit

File: yt_dlp/extractor/laracasts.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+        metadata = {
+            'display_id': display_id,
+            **traverse_obj(series, {
+                'title': ('title', {str}),
+                'id': ('id', {int}, {str_or_none}),
+                'description': ('body', {clean_html}),
+                'thumbnail': (('large_thumbnail', 'thumbnail'), any, {url_or_none}),
+                'duration': ('runTime', {parse_duration}),
+                'categories': ('taxonomy', 'name', {lambda x: x and [x]}),
```
Comment (seproDev) [inline yt_dlp/extractor/laracasts.py:110]:
```suggestion
                'categories': ('taxonomy', 'name', {str}, {lambda x: x and [x]}),
```

File: yt_dlp/extractor/laracasts.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            'tags': ['Laravel'],
+            'display_id': '30-days-to-learn-laravel-11',
+        },
+        'playlist_count': 30,
+    }]
+
+    def _entries(self, series, episode_count):
+        for current_episode in range(1, episode_count + 1):
+            webpage_url = f'https://laracasts.com/series/{series}/episodes/{current_episode}'
+            yield self.url_result(webpage_url, LaracastsIE)
```
Comment (bashonly) [inline yt_dlp/extractor/laracasts.py:96]: this method is never used?

File: yt_dlp/extractor/laracasts.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+from ..utils.traversal import traverse_obj
+
+
+class LaracastsBaseIE(InfoExtractor):
+    def get_prop_data(self, url, display_id):
+        webpage = self._download_webpage(url, display_id)
+        return traverse_obj(
+            get_element_html_by_id('app', webpage), ({extract_attributes}, 'data-page', {json.loads}, 'props'))
+
+    def parse_episode(self, episode):
```
Comment (bashonly) [inline yt_dlp/extractor/laracasts.py:25]: private methods
```suggestion
    def _get_prop_data(self, url, display_id):
        webpage = self._download_webpage(url, display_id)
        return traverse_obj(
            get_element_html_by_id('app', webpage), ({extract_attributes}, 'data-page', {json.loads}, 'props'))

    def _parse_episode(self, episode):
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/laracasts.py (hunk 5 above)
Comment (seproDev) [reply yt_dlp/extractor/laracasts.py:96]: Ah, missed removing that during refactoring

File: yt_dlp/extractor/laracasts.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    def _get_prop_data(self, url, display_id):
+        webpage = self._download_webpage(url, display_id)
+        return traverse_obj(
+            get_element_html_by_id('app', webpage), ({extract_attributes}, 'data-page', {json.loads}, 'props'))
+
+    def _parse_episode(self, episode):
+        if not episode.get('vimeoId'):
+            self.raise_login_required('This video is only available for subscribers.')
+        return self.url_result(
+            VimeoIE._smuggle_referrer(f'https://player.vimeo.com/video/{episode["vimeoId"]}', 'https://laracasts.com/'),
```
Comment (bashonly) [inline yt_dlp/extractor/laracasts.py:29]:
```suggestion
            VimeoIE._smuggle_referrer(
                f'https://player.vimeo.com/video/{episode["vimeoId"]}', 'https://laracasts.com/'),
```

File: yt_dlp/extractor/laracasts.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+
+class LaracastsBaseIE(InfoExtractor):
+    def _get_prop_data(self, url, display_id):
+        webpage = self._download_webpage(url, display_id)
+        return traverse_obj(
+            get_element_html_by_id('app', webpage), ({extract_attributes}, 'data-page', {json.loads}, 'props'))
+
+    def _parse_episode(self, episode):
+        if not episode.get('vimeoId'):
```
Comment (bashonly) [inline yt_dlp/extractor/laracasts.py:26]:
```suggestion
        if not traverse_obj(episode, 'vimeoId'):
```

File: yt_dlp/extractor/laracasts.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    urljoin,
+)
+from ..utils.traversal import traverse_obj
+
+
+class LaracastsBaseIE(InfoExtractor):
+    def _get_prop_data(self, url, display_id):
+        webpage = self._download_webpage(url, display_id)
+        return traverse_obj(
+            get_element_html_by_id('app', webpage), ({extract_attributes}, 'data-page', {json.loads}, 'props'))
```
Comment (bashonly) [inline yt_dlp/extractor/laracasts.py:23]:
```suggestion
            get_element_html_by_id('app', webpage),
            ({extract_attributes}, 'data-page', {json.loads}, 'props'))
```

File: yt_dlp/extractor/laracasts.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            self.raise_login_required('This video is only available for subscribers.')
+        return self.url_result(
+            VimeoIE._smuggle_referrer(f'https://player.vimeo.com/video/{episode["vimeoId"]}', 'https://laracasts.com/'),
+            VimeoIE, url_transparent=True,
+            **traverse_obj(episode, {
+                'id': ('id', {int}, {str_or_none}),
+                'webpage_url': ('path', {lambda x: urljoin('https://laracasts.com', x)}),
+                'title': ('title', {clean_html}),
+                'season_number': ('chapter', {int}),
+                'episode_number': ('position', {int}),
```
Comment (bashonly) [inline yt_dlp/extractor/laracasts.py:36]:
```suggestion
                'season_number': ('chapter', {int_or_none}),
                'episode_number': ('position', {int_or_none}),
```

File: yt_dlp/extractor/laracasts.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                'description': ('body', {clean_html}),
+                'thumbnail': ('largeThumbnail', {url_or_none}),
+                'duration': ('length', {int_or_none}),
+                'date': ('dateSegments', 'published', {unified_strdate}),
+            }))
+
+
+class LaracastsIE(LaracastsBaseIE):
+    IE_NAME = 'laracasts'
+    _VALID_URL = r'https?://(?:www\.)?laracasts\.com/series/(?P<id>[\w\d-]+/episodes/[0-9]+)/?(?:[?#]|$)'
```
Comment (bashonly) [inline yt_dlp/extractor/laracasts.py:46]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?laracasts\.com/series/(?P<id>[\w-]+/episodes/\d+)/?(?:[?#]|$)'
```

File: yt_dlp/extractor/laracasts.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    }]
+
+    def _real_extract(self, url):
+        display_id = self._match_id(url)
+        return self._parse_episode(self._get_prop_data(url, display_id)['lesson'])
+
+
+class LaracastsPlaylistIE(LaracastsBaseIE):
+    IE_NAME = 'laracasts:series'
+    _VALID_URL = r'https?://(?:www\.)?laracasts\.com/series/(?P<id>[\w\d-]+)/?(?:[?#]|$)'
```
Comment (bashonly) [inline yt_dlp/extractor/laracasts.py:76]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?laracasts\.com/series/(?P<id>[\w-]+)/?(?:[?#]|$)'
```

**Review by bashonly — APPROVED** [review]

**Comment by bashonly** [comment]
@dangnhdev please open a new "Site feature request" issue

**Comment by seproDev** [comment]
@dangnhdev It actually already being extracted as `season_number`. Still, merged PRs are not the place to discuss this. Open a new issue if you need help.

**Pull request merged**